crates/trie/sparse/src/trie.rs

```diff
@@ -715,23 +790,95 @@ impl RevealedSparseTrie {
                             continue 'main
                         }
                     }
+                    hashes.reverse();
 
                     self.rlp_buf.clear();
-                    let rlp_node = BranchNodeRef::new(&buffers.branch_value_stack_buf, *state_mask)
-                        .rlp(&mut self.rlp_buf);
+                    let branch_node_ref =
+                        BranchNodeRef::new(&buffers.branch_value_stack_buf, *state_mask);
+                    let rlp_node = branch_node_ref.rlp(&mut self.rlp_buf);
                     *hash = rlp_node.as_hash();
-                    rlp_node
+
+                    let mut tree_mask_values = tree_mask_values.into_iter().rev();
+                    let mut hash_mask_values = hash_mask_values.into_iter().rev();
+                    let mut tree_mask = TrieMask::default();
+                    let mut hash_mask = TrieMask::default();
+                    for (i, child) in branch_node_ref.children() {
+                        if child.is_some() {
+                            if hash_mask_values.next().unwrap() {
+                                hash_mask.set_bit(i);
+                            }
+                            if tree_mask_values.next().unwrap() {
+                                tree_mask.set_bit(i);
+                            }
+                        }
+                    }
+
+                    // Store in DB trie if there are either any children that are stored in the DB
+                    // trie, or any children represent hashed values
+                    let store_in_db_trie = !tree_mask.is_empty() || !hash_mask.is_empty();
```

Comment:
all of this is necessary only to determine whether we need to store the branch node, so worth bringing it under `if let Some(updates) = self.updates.as_mut() {`